### PR TITLE
Evolve feature PR

### PR DESCRIPTION
**Pull request by ikarabulut**
@riccjohn Here is my updated PR with the evolve-feature to start the loop through worlds.

### CODE REVIEW

**Comment by riccjohn** [comment]
We can probably add a `.gitignore` file to ignore `.irb_history`

**Comment by riccjohn** [comment]
Overall comment: Right now we have 3 ways of notating a cells' alive/dead status:
- A `Cell` internally has a `status` of `1` or `0`
- A `Cell` responds to `alive?` and `dead?` with booleans
- `World` converts dead cells to `o` and live cells to `x`

I think the only time we should convert an instance of a `Cell` into an `"x"`/`"o"` or `1`/`0` is when we're displaying it on screen. The actual manipulation of cells should all be done by the `Cell` class (or  more specifically, the `World` sending messages to the `Cell` class).

We might want to start by having a `Cell` get initialized with a symbol, like `:alive` or `:dead`, and having all of the `World` class deal with that status, and only the display-portion deals with converting that to "x"/"o", or however we want to display it.

File: lib/cell.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+require_relative './world.rb'
+
+class Cell
+  attr_reader :x, :y
+  attr_accessor :status
```
Comment (ikarabulut) [inline lib/cell.rb:5]: No need for accessor since we have #die and #revive

File: lib/world.rb [hunk 2]
```diff
@@ -104,4 +108,48 @@ def revive_at?(x, y)
...
+      puts "Generation #{generation_number += 1}"
+      tick
+      pp display_board
+      sleep(1)
+    end
+  end
+
+  def tick
+    old_world = cell_objects
+    new_world = Array.new(@rows) { Array.new(@columns) }
```
Comment (ikarabulut) [inline lib/world.rb:124]: Could be a method?

File: lib/world.rb [hunk 3]
```diff
@@ -104,4 +108,48 @@ def revive_at?(x, y)
...
+    old_world.each do |cell|
+      new_cell = Cell.new(cell.x, cell.y)
+      cell = get_cell(cell.x, cell.y)
+      if cell.dead?
+        if revive_at?(cell.x, cell.y)
+          new_cell.status = 1
```
Comment (ikarabulut) [inline lib/world.rb:130]: Use the revive method

File: lib/world.rb [hunk 4]
```diff
@@ -104,4 +108,48 @@ def revive_at?(x, y)
...
+          new_world[cell.x][cell.y] = new_cell
+        end
+      elsif cell.alive?
+        if alive_next_generation?(cell.x, cell.y)
+          new_cell.status = 1
```
Comment (ikarabulut) [inline lib/world.rb:135]: use revive method

File: lib/world.rb [hunk 5]
```diff
@@ -104,4 +108,48 @@ def revive_at?(x, y)
...
+          new_world[cell.x][cell.y] = new_cell
+        end
+      else
```
Comment (ikarabulut) [inline lib/world.rb:138]: delete

**Review by riccjohn — CHANGES_REQUESTED** [review]

File: lib/program_run.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+require_relative './world.rb'
```
[the remark this answers is not in the document]
Comment (jramnani) [reply lib/program_run.rb:1]: Yup! That's a good place for it. 👍

File: lib/world.rb [hunk 7]
```diff
@@ -58,35 +65,32 @@ def neighbors_of(x, y)
...
+    return neighbors
   end
 
   def alive_neighbors_of(x, y)
     alive_neighbors = []
-    neighbors_of(x, y).each do |neighbor|
-      if neighbor == "x"
+    neighbors = neighbors_of(x, y)
+
+    neighbors.each do |neighbor|
```
Comment (jramnani) [inline lib/world.rb:87]: Now that you have a message you can send a Cell to tell whether it's alive, you are able to simplify this loop using some more advanced Ruby.

```
alive_neighbors = neighbors.filter { |neighbor| neighbor.alive? }
```

Comment (ikarabulut) [reply lib/world.rb:87]: New efficient code structures excite me

**Pull request closed without merging**